packages/dataviews/src/dataform-hooks/use-form.ts

```diff
@@ -0,0 +1,49 @@
+/**
+ * WordPress dependencies
+ */
+import { useState } from '@wordpress/element';
+
+/**
+ * Internal dependencies
+ */
+import type { FormField } from '../types';
+
+export const useForm = < Item >(
+	supportedFields: Record< string, FormField >
+) => {
+	const [ form, setForm ] = useState( {
+		fields: supportedFields,
+		touchedFields: [] as string[],
+		messageErrors: {},
+	} );
+
+	const setTouchedFields = ( touchedFields: string[] ) => {
+		setForm( ( prevForm ) => ( {
+			...prevForm,
+			touchedFields,
+		} ) );
+	};
+
+	const setErrors = ( field: string, error: string | undefined ) => {
+		setForm( ( prevForm ) => ( {
+			...prevForm,
+			messageErrors: {
+				...form.messageErrors,
+				[ field ]: error,
+			},
+		} ) );
+	};
+
+	const isFormValid = ( data: Item ) => {
+		return Object.entries( form.fields ).every( ( [ , field ] ) => {
+			return field.validation.callback( data ).isValid;
+		} );
+	};
+
+	return {
+		...form,
+		setTouchedFields,
+		setErrors,
+		isFormValid,
+	};
```

Comment:
Some of these are used as dependencies in hooks later... so probably better to memoize (useEvent/useCallback)